source/common/http/conn_manager_impl.cc

```diff
@@ -1467,6 +1468,29 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ResponseHeaderMap& heade
 
   chargeStats(headers);
 
+  if (auto* proxy_status_config = connection_manager_.config_.proxyStatusConfig();
+      proxy_status_config != nullptr) {
+    // Writing the Proxy-Status header is gated on the existence of
+    // |proxy_status_config|. The |details| field and other internals are generated in
+    // fromStreamInfo().
+    if (absl::optional<StreamInfo::ProxyStatusError> proxy_status =
+            StreamInfo::ProxyStatusUtils::fromStreamInfo(filter_manager_.streamInfo());
+        proxy_status.has_value()) {
+      headers.appendProxyStatus(StreamInfo::ProxyStatusUtils::toString(
```

Comment:
I did not find a test for multiple Proxy-Status headers. You may need one with upstream including Proxy-Status in its response.